src/util/timer.hpp

```diff
@@ -1,170 +1,58 @@
 #pragma once
 
 #include <string>
-#include <vector>
-#include <utility>
-#include <map>
 #include <chrono>
-#include <numeric>
-#include <algorithm>
-#include <fstream>
-#include <functional>
-
 #include <json.hpp>
-#include "filesystem.hpp"
-
 
 /// A simple stack-based library used to time nested parts of a program.
-///
-/// if !OTTO_ENABLE_TIMERS, all implementations are empty
 namespace otto::util::timer {
-
-  /// The type used to identify timers
   using timer_id = std::string;
+  using time_point = std::chrono::time_point<
+    std::chrono::steady_clock, std::chrono::nanoseconds>;
+  using time_duration = time_point::duration::rep;
 
-  /// Contains the actual timing data.
-  ///
-  /// Should only very rarely be interacted with directly by the user.
-  struct Timer {
-
-    /// A `std::chrono::time_point` used to measure timings.
-    using time_point = std::chrono::time_point<
-      std::chrono::steady_clock, std::chrono::nanoseconds>;
-
-    /// Construct a new timer with id `id`
-    ///
-    /// \effects set member variable `id` to `id`
-    /// \remarks at this point, `start_time` is undefined.
-    Timer(const timer_id& id);
-
-    /// Construct a new timer with id `id`
-    ///
-    /// \effects set member variable `id` to `id`
-    /// \remarks at this point, `start_time` is undefined.
-    Timer(timer_id&& id);
-
-    /// Start the timer
-    ///
-    /// \effects sets `start_time` to `time_point::clock::now()`
-    /// \postconditions `running == true`
-    /// \remarks Do *not* use this function. In most cases you want the free
-    /// function which also pushes and pops from/to the stack
-    void start();
-
-    /// Stop the timer
-    ///
-    /// \requires `running == true`
-    /// \effects pushes `time_point::clock::now() - start_time` to the back of `data`.
-    /// \postconditions `running == false`
-    /// \remarks Do *not* use this function. In most cases you want the free
-    /// function which also pushes and pops from/to the stack
-    void stop();
-
-    /// Tick the timer
-    ///
-    /// \effects if `running == false`, same as `start()`. Otherwise, push
-    /// `time_point::clock::now() - start_time` to the back of `data`, and
-    /// assign `time_point::clock::now()` to `start_time`
-    /// \postconditions `running == true`
-    /// \complexity `time_point::clock::now()` is only called once.
-    /// \remarks Do *not* use this function. In most cases you want the free
-    /// function which also pushes and pops from/to the stack
-    void tick();
```

Comment:
we still need an interface for "ticking", and in general manual (non-scoped) timer handling. However to make sure this is not used by default by accident, maybe we can name the functions `timer::manual_start()` etc?

Reply:
What is the purpose for a "ticking" interface?  If the purpose of the timing data is to measure comparative performance, the data that is gathered via a scoped interface should be more than sufficient.

Take a look at the serialized data in timers.json -- is there a specific scenario you'd *currently* like to handle that isn't captured by this data?

Reply:
the main thing is audio processing. I would like to both time the actual sample rate, and how much of that time we actually spend in the `process` function. To do that, we could tick a timer at the beginning of each audio callback, which should get us the sample rate, and use a scopetimer to time the scope, which will give us the time we spent there. The main reason not to just calculate it using the sample rate, is i want to have some python scripts to visualize this data, and not have to handle audio as a special case

Reply:
Would you be willing to accept this change as-is for now, and we can separately revisit that issue in its own PR?

Reply:
definately, was just about to say so